presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -530,39 +660,111 @@ public boolean equals(Object obj)
+                for (VariableReferenceExpression variable : assignments.keySet()) {
+                    resolveVariable(variable, resolvedVariables, assignments, availableVariables);
+                }
+
+                return Assignments.builder().putAll(assignments).build();
+            }
+
+            private void resolveVariable(VariableReferenceExpression variable, HashSet<VariableReferenceExpression> resolvedVariables, Map<VariableReferenceExpression,
+                    RowExpression> assignments, Set<VariableReferenceExpression> availableVariables)
+            {
+                RowExpression expression = assignments.get(variable);
+
+                Sets.SetView<VariableReferenceExpression> variablesToResolve = Sets.difference(Sets.difference(extractUnique(expression), availableVariables), resolvedVariables);
+                if (variablesToResolve.isEmpty()) {
+                    resolvedVariables.add(variable);
+                    return;
+                }
+
+                variablesToResolve.forEach(variableToResolve -> resolveVariable(variableToResolve, resolvedVariables, assignments, availableVariables));
+                // Modify the assignments to replace the variables with the resolved expressions
+                assignments.put(variable, replaceExpression(expression, assignments));
+                // Mark the variable as resolved
+                resolvedVariables.add(variable);
             }
 
-            private void flattenNode(PlanNode node, int limit)
+            private void rewriteFilterWithInlinedAssignments(Assignments assignments)
+            {
+                ImmutableList.Builder<RowExpression> modifiedFilters = ImmutableList.builder();
+                filters.forEach(filter -> modifiedFilters.add(replaceExpression(filter, assignments.getMap())));
+                filters = modifiedFilters.build();
+            }
+
+            private void flattenNode(PlanNode node, int limit, Map<VariableReferenceExpression, RowExpression> assignmentsBuilder)
             {
                 PlanNode resolved = lookup.resolve(node);
 
+                if (resolved instanceof ProjectNode) {
+                    ProjectNode projectNode = (ProjectNode) resolved;
+                    // A ProjectNode could be 'hiding' a join source by building an assignment of a complex equi-join criteria like `left.key = right1.key1 + right1.key2`
+                    // We open up the join space by tracking the assignments from this Project node; these will be inlined into the overall filters once we finish
+                    // traversing the join graph
+                    if (handleComplexEquiJoins && lookup.resolve(projectNode.getSource()) instanceof JoinNode) {
+                        //We track only the non-identity assignments since these are the ones that will be inlined into the overall filters
+                        assignmentsBuilder.putAll(getNonIdentityAssignments(projectNode.getAssignments()));
```

Comment:
Should we exclude non deterministic assignments here?
Also, if the assignment is a try() statement, will this optimization changes the semantic?

Reply:
Good catch! I think I can enhance this more to work with non deterministic assignments (and in general ReorderJoins to work with non-deterministic filters), but for now, I've made the change to only consider flattening the Project node if none of its assignments are non-deterministic.  
I've added a `testProjectNodesWithNonDeterministicAssignmentsAreNotFlattenedForComplexEquiJoins` to assert this behavior

Reply:
Since TRY is a non-deterministic, we will not flatten a Project with it